src/util/simplify_expr2.cpp

```diff
@@ -415,6 +415,258 @@ static type2tc common_arith_op2_type(expr2tc &e, expr2tc &f)
 
 expr2tc add2t::do_simplify() const
 {
+  // X + X --> X << 1
+  if(is_symbol2t(side_1) && is_symbol2t(side_2))
+  {
+    if(to_symbol2t(side_1) == to_symbol2t(side_2))
```

Comment:
I don't think you need to type-cast in order to compare expressions. Isn't `side_1 == side_2` enough?